File: uploader/src/main/java/net/hearthstats/Deck.java
package net.hearthstats;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import lombok.Value;
import lombok.experimental.Accessors;
import lombok.experimental.Builder;

import org.json.simple.JSONObject;

@Value
@Builder
@Accessors(fluent = true)
public class Deck {
	private int id;
	private String name;
	private List<Card> cards;

	public boolean isValid() {
		if (cards != null) {
			int count = 0;
			for (Card c : cards) {
				count += c.count();
			}
			return count == 30;
		} else
			return false;
	}

	public static Deck fromJson(JSONObject json) {
		int id = Integer.parseInt(json.get("id").toString());
		Object string = json.get("cardstring");

		List<Card> cardList = null;
		if (string != null) {
			String cardString = string.toString().trim();
			if (cardString.length() > 0) {
				cardList = parseDeckString(cardString);
			}
		}
		return new Deck(id, json.get("name").toString(), cardList);

	}

	private static List<Card> parseDeckString(String ds) {
		Map<Integer, Card> cardData = CardUtils.getCards();
		List<Card> cards = new ArrayList<>();
		for (String card : ds.split(",")) {
			int u = card.indexOf('_');
			String count = card.substring(u + 1);
			int id = Integer.parseInt(card.substring(0, u));
			Card cd = cardData.get(id);
			cards.add(Card.builder().id(id).name(cd.name()).cost(cd.cost())
					.count(Integer.parseInt(count)).build());
		}
		Collections.sort(cards);
		return cards;
	}
}

File: uploader/src/main/scala/net/hearthstats/Deck.scala
package net.hearthstats

import java.text.ParseException

import org.apache.commons.lang3.StringUtils
import org.json.simple.JSONObject

import scala.collection.JavaConversions._

case class Deck(
  id: Int = -1,
  name: String = "",
  slug: String = "",
  cards: List[Card] = List.empty,
  hero: String = "Shaman",
  activeSlot: Option[Int] = None) {

  def isValid =
    cards != null && 30 == cardCount

  def cardCount =
    cards.map(_.count).sum

  def klassId: Int = {
    (0 until Constants.hsClassOptions.length).find(Constants.hsClassOptions(_) == hero).getOrElse(0)
  }
  
  def cardString: String = {
    (for (c <- cards.sortBy(_.id)) yield s"${c.id}_${c.count}").mkString(",")
  }

  def deckString: String = {
    (for (c <- cards.sortBy(card => (card.cost, card.name)))
      yield s"${c.count} ${c.originalName}").mkString("\n")
  }

  def toJsonObject: JSONObject = {
    return new JSONObject(collection.mutable.Map(
      "klass_id" -> klassId,
      "name" -> name,
      "notes" -> "",
      "cardstring" -> cardString
    ))
  }

  override def toString = s"$hero: $name"
}

object Deck {

  def fromJson(json: JSONObject): Deck = {
    val id = Integer.parseInt(json.get("id").toString)
    val cardList =
      Option(json.get("cardstring")) match {
        case Some(cs) if StringUtils.isNotBlank(cs.toString) =>
          parseCardString(cs.toString.trim)
        case _ => Nil
      }

    val klassId = json.get("klass_id")
    val heroString = if (klassId == null) "" else Constants.hsClassOptions(klassId.toString.toInt)

    Deck(id = id,
      slug = json.get("slug").toString,
      name = json.get("name").toString,
      cards = cardList,
      hero = heroString,
      activeSlot = Option(json.get("slot")).map(_.toString.toInt))
  }

  def parseCardString(ds: String): List[Card] = {
    val cardData = CardUtils.cards
    val cards = for {
      card <- ds.split(",").toList
      u = card.indexOf('_')
      count = card.substring(u + 1)
      id = Integer.parseInt(card.substring(0, u))
      cd = cardData(id)
    } yield cd.copy(count = Integer.parseInt(count))
    cards.sorted
  }

  def parseDeckString(ds: String): List[Card] = {
    val cardMap = CardUtils.cards.values.map(c => c.name -> c).toMap
    val cards = for {
      card <- ds.split("\n").toList
      count = try {
        Integer.parseInt(card.substring(0, 1))
      } catch {
        case e: NumberFormatException => 0
      }
      cd = cardMap.get(card.substring(2))
    } yield cd match {
        case Some(c) =>
          c.copy(count = count)
        case None =>
          // If the card is invalid, then return a made-up card. This might be better implemented as a different class
          // to indicate that the card couldn't be identified
          new Card(id = 0, originalName = card, collectible = false)
      }
    cards.sorted
  }
}

